crates/holochain/src/test_utils/consistency.rs

```diff
@@ -0,0 +1,479 @@
+    missing_hashes: &mut Vec<Arc<KitsuneOpHash>>,
+) -> DatabaseResult<()> {
+    missing_hashes.clear();
+    // We need to swap these hashes so we can move them into the async_reader
+    // without reallocating.
+    let expected = std::mem::replace(expected_hashes, Vec::with_capacity(0));
+    let mut missing = std::mem::replace(missing_hashes, Vec::with_capacity(0));
+    let mut r = vault
+                .async_reader(move |txn| {
+                    for hash in &expected {
+                        // TODO: This might be too slow, could instead save the holochain hash versions.
+                        let h_hash: DhtOpHash = DhtOpHashExt::from_kitsune(&hash);
+                        let integrated: bool = txn.query_row(
+                            "
+                            SELECT EXISTS(
+                                SELECT 1 FROM DhtOp WHERE hash = :hash AND when_integrated IS NOT NULL
+                            )
+                            ",
+                            named_params! {
+                                ":hash": h_hash,
+                            },
+                            |row| row.get(0),
+                        )?;
+                        if !integrated {
+                            missing.push(hash.clone());
+                        }
+                    }
+                    DatabaseResult::Ok((expected, missing))
+                })
+                .await?;
+    // Put the data back.
+    std::mem::swap(&mut r.0, expected_hashes);
+    std::mem::swap(&mut r.1, missing_hashes);
+    Ok(())
+}
+
+async fn gather_published_data(
+    iter: impl Iterator<Item = Stores>,
+    concurrency: usize,
+) -> StateQueryResult<Vec<PublishedData>> {
+    use futures::stream::TryStreamExt;
+    let iter = iter.map(|stores| async move {
+        let published_hashes = request_published_ops(&stores.cell_env).await?;
+        let storage_arc = request_arc(&stores.p2p_env, &stores.agent).await?;
+        Ok(storage_arc.map(|storage_arc| PublishedData {
+            agent: stores.agent,
+            storage_arc,
+            published_hashes,
+        }))
+    });
+    futures::stream::iter(iter)
+        .buffer_unordered(concurrency)
+        .try_filter_map(|d| futures::future::ok(d))
+        .try_collect()
+        .await
+}
+
+async fn request_published_ops(env: &EnvRead) -> StateQueryResult<Vec<Arc<KitsuneOpHash>>> {
+    Ok(env
+        .async_reader(|txn| {
+            // Collect all ops except StoreEntry's that are private.
+            let r = txn
+                .prepare(
+                    "
+                    SELECT
+                    DhtOp.hash as dht_op_hash
+                    FROM DhtOp
+                    JOIN
+                    Header ON DhtOp.header_hash = Header.hash
+                    WHERE
+                    DhtOp.is_authored = 1
+                    AND
+                    (DhtOp.type != :store_entry OR Header.private_entry = 0)
```

Comment:
Just curious, is the DhtOp.type check redundant? Can something be private that's not a StoreEntry?

Reply:
Yes but we publish private ops that aren't the StoreEntry